crates/typst-library/src/math/matrix.rs

```diff
@@ -232,55 +302,156 @@ fn layout_vec_body(
         }
     }
     ctx.unstyle();
 
-    let mut frame = Frame::new(Size::new(
-        Abs::zero(),
-        heights.iter().map(|&(a, b)| a + b).sum::<Abs>() + row_gap * (nrows - 1) as f64,
-    ));
+    // For each row, combine maximum ascent and descent into a row height.
+    // Sum the row heights and then add the total height of the gaps between rows.
+    let total_height =
+        heights.iter().map(|&(a, b)| a + b).sum::<Abs>() + row_gap * (nrows - 1) as f64;
+
+    // Width starts at zero because it can't be calculated until later
+    let mut frame = Frame::new(Size::new(Abs::zero(), total_height));
+
     let mut x = Abs::zero();
-    for col in cols {
+
+    for (index, col) in cols.into_iter().enumerate() {
         let AlignmentResult { points, width: rcol } = alignments(&col);
+
         let mut y = Abs::zero();
+
         for (cell, &(ascent, descent)) in col.into_iter().zip(&heights) {
             let cell = cell.into_aligned_frame(ctx, &points, Align::Center);
             let pos = Point::new(
                 if points.is_empty() { x + (rcol - cell.width()) / 2.0 } else { x },
                 y + ascent - cell.ascent(),
             );
+
             frame.push_frame(pos, cell);
+
             y += ascent + descent + row_gap;
         }
-        x += rcol + col_gap;
+
+        // Advance to the end of the column
+        x += rcol;
+
+        // If a vertical line should be inserted after this column
+        if vline.contains(&(index + 1)) {
+            frame.push(
+                Point::with_x(x + half_col_gap),
+                vline_item(total_height, stroke.clone(), span),
+            );
+        }
+
+        // Advance to the start of the next column
+        x += col_gap;
+    }
+
+    // Once all the columns are laid out, the total width can be calculated
+    let total_width = x - col_gap;
+
+    // This allows the horizontal lines to be laid out
+    for line in hline {
+        let offset = (heights[0..line].iter().map(|&(a, b)| a + b).sum::<Abs>()
+            + row_gap * (line - 1) as f64)
+            + half_row_gap;
+
+        frame.push(Point::with_y(offset), hline_item(total_width, stroke.clone(), span));
     }
-    frame.size_mut().x = x - col_gap;
+
+    frame.size_mut().x = total_width;
 
     Ok(frame)
 }
 
-/// Layout the outer wrapper around a vector's or matrices' body.
+fn hline_item(length: Abs, stroke: Stroke, span: Span) -> FrameItem {
+    let hline_geom = Geometry::Line(Point::with_x(length));
+
+    FrameItem::Shape(
+        Shape {
+            geometry: hline_geom,
+            fill: None,
+            stroke: Some(stroke),
+        },
+        span,
+    )
+}
+
+fn vline_item(length: Abs, stroke: Stroke, span: Span) -> FrameItem {
+    let vline_geom = Geometry::Line(Point::with_y(length));
+
+    FrameItem::Shape(
+        Shape {
+            geometry: vline_geom,
+            fill: None,
+            stroke: Some(stroke),
+        },
+        span,
+    )
+}
```

Comment:
these two functions are pretty similar, wouldn't it be easier to pass in the point and then have it as one function?